mcp/streamable.go

```diff
@@ -472,58 +480,45 @@ func (c *streamableServerConn) SessionID() string {
 // at any time.
 type stream struct {
 	// id is the logical ID for the stream, unique within a session.
-	// an empty string is used for messages that don't correlate with an incoming request.
+	//
+	// The standalone SSE stream has id "".
 	id string
 
-	// If isInitialize is set, the stream is in response to an initialize request,
-	// and therefore should include the session ID header.
-	isInitialize bool
-
-	// jsonResponse records whether this stream should respond with application/json
-	// instead of text/event-stream.
-	//
-	// See [StreamableServerTransportOptions.JSONResponse].
-	jsonResponse bool
+	// mu guards the fields below, as well as storage of new messages in the
+	// connection's event store (if any).
+	mu sync.Mutex
 
-	// signal is a 1-buffered channel, owned by an incoming HTTP request, that signals
-	// that there are messages available to write into the HTTP response.
-	// In addition, the presence of a channel guarantees that at most one HTTP response
-	// can receive messages for a logical stream. After claiming the stream, incoming
-	// requests should read from the event store, to ensure that no new messages are missed.
-	//
-	// To simplify locking, signal is an atomic. We need an atomic.Pointer, because
-	// you can't set an atomic.Value to nil.
+	// If non-nil, deliver writes data directly to the HTTP response.
 	//
-	// Lifecycle: each channel value persists for the duration of an HTTP POST or
-	// GET request for the given streamID.
-	signal atomic.Pointer[chan struct{}]
-
-	// The following mutable fields are protected by the mutex of the containing
-	// StreamableServerTransport.
+	// Only one HTTP response may receive messages at a given time. An active
+	// HTTP connection acquires ownership of the stream by setting
+	deliver func(data []byte, final bool) error
 
-	// streamRequests is the set of unanswered incoming RPCs for the stream.
+	// streamRequests is the set of unanswered incoming requests for the stream.
 	//
-	// Requests persist until their response data has been added to the event store.
+	// Requests are removed when their response has been received.
 	requests map[jsonrpc.ID]struct{}
 }
 
-func (c *streamableServerConn) newStream(ctx context.Context, id string, isInitialize, jsonResponse bool) (*stream, error) {
-	if err := c.eventStore.Open(ctx, c.sessionID, id); err != nil {
-		return nil, err
+// done reports whether the stream is logically complete.
+//
+// s.mu must be held while calling this function.
+func (s *stream) done() bool {
```

Comment:
This is your codebase, so use whatever local style you prefer, but I'd call this `doneLocked` to indicate the stream mutex must be held.

Reply:
+1

Reply:
Funny, I normally do that but thought it was specific to *my* style, so opted not to do it here.